### PR TITLE
Minimum execution time for `withRootKey` on authentication failure.

### PR DESCRIPTION
**Pull request by jonathanknowles**
# Issue Number

#1984 

# Overview

This PR adjusts the `withRootKey` function to incorporate a minimum execution time in the event of failure.

- In the event of a _successful_ authentication, there is _no_ delay.
- In the event of an _unsuccessful_ authentication, the total execution time of the `withRootKey` function is:
    -  equal to (or greater than) `minimumExecutionTimeOnAuthFailure`.
    -  independent of the actual time required to perform the authentication.

Where `minimumExecutionTimeOnAuthFailure` is defined as:
```hs
-- | Specifies a minimum bound on the length of time that is allowed to elapse
--   before an authentication failure can be reported to the user.
--
minimumExecutionTimeOnAuthFailure :: NominalDiffTime   
minimumExecutionTimeOnAuthFailure = threeSeconds   
  where   
    threeSeconds = fromRational 3   
```

In the future, we could make this configurable (if the need arises).

# Testing

- [x] Add integration test to confirm that an unsuccessful attempt to authenticate results in a delay of `minimumExecutionTimeOnAuthFailure` or longer.
- [x] Add integration test to confirm that a successful attempt to authenticate does not result in such a delay.

### CODE REVIEW

**Comment by jonathanknowles** [comment]
bors try

**Comment by jonathanknowles** [comment]
bors try

**Comment by jonathanknowles** [comment]
bors try

File: lib/core/src/Data/Time/Utils.hs [hunk 1]
```diff
@@ -23,3 +37,16 @@ utcTimeSucc = addUTCTime $ succ 0
...
+-- | Suspends the current thread until after the specified time has passed.
+--
+-- There is no guarantee that the thread will be rescheduled promptly when the
+-- delay has expired, but the thread will never continue to run earlier than
+-- specified.
+--
+waitUntil :: MonadIO m => UTCTime -> m ()
+waitUntil futureTime = liftIO $ do
+    currentTime <- getCurrentTime
+    threadDelay
```
Comment (rvl) [inline lib/core/src/Data/Time/Utils.hs:50]: Does this work OK when the difference is negative?
Perhaps you could put a `max 0` there to make sure.

**Review by rvl — COMMENTED** [review]
It looks OK at first glance if we are going to implement this.

File: lib/core/src/Data/Time/Utils.hs (hunk 1 above)
Comment (jonathanknowles) [reply lib/core/src/Data/Time/Utils.hs:50]: > Does this work OK when the difference is negative?
> Perhaps you could put a `max 0` there to make sure.

I think it should work, as informal testing of `threadDelay` shows that it exits immediately when called with a negative value.

However, this behaviour appears to be undocumented, and so I agree it's a good idea to use `max 0`, as then we won't have to rely on undocumented behaviour.

Fixed in bcdeeb8d8327d9b91aacbd6ec71e5545547c4989.

**Comment by KtorZ** [comment]
Closing for now. Will re-open once we've got feedback from the security auditors.

**Pull request closed without merging**